### PR TITLE
Use schema provider to remove Catalog dependency from QueryEngine

### PR DESCRIPTION
**Pull request by alexbaden**
I changed schema provider to return an ID matching the catalog database ID, as each Catalog is locked to a single database. If we want to keep that change going forward, we can consider removing database ID as a separate param entirely from Executor. 

Still a few issues to work out but opening this for discussion of the above, and any other issues that might exist. Note that there is still Catalog dependence in some C++ files, but the public API no longer uses catalog.

### CODE REVIEW

**Comment by ienkovich** [comment]
The idea was to have a `SchemaMgr` that implements `SchemaProvider` interface and gives access to multiple data sources. It would hold multiple `SchemaProviders` and make dispatching by database id. The highest byte of database id is to be used as schema id to choose a proper `SchemaProvider`. getId method of `SchemaProvider` is to return this schema id for proper dispatching, not database id. `DataProviders` should work in a similar way (`PersistentStorageMgr` is used as such dispatcher for them). We need this to work in Presto worker where we might have multiple data sources mixed in a single query and therefore might have multiple data/schema providers used for a single query processing.

Currently, as a temporary fix to not access the `Catalog`, `RelAlgExecutor` and `Executor` have `db_id_` field and [get|set]DatabaseId methods. In my plan, I have a task to remove this restriction and allow multiple database ids in a single query (mostly by holding full IDs including db_id for tables, dicts, etc. in various structures).

If we need to get database id from `SchemaProvider` right now to get rid of `Catalog`, we can temporarily add `getDatabaseId()` method to `SchemaProvider` interface. In the implementation, we would check that there is a single database available in the provider and return its id. But looking at your current patch, it looks like `Executor::getDatabaseId` can be used in all cases where you use `SchemaProvider::getId`.

File: QueryEngine/RelAlgExecutor.cpp [hunk 1]
```diff
@@ -117,7 +117,6 @@ size_t RelAlgExecutor::getOuterFragmentCount(const CompilationOptions& co,
   ScopeGuard row_set_holder = [this] { cleanupPostExecution(); };
   const auto col_descs = get_physical_inputs(&ra);
   const auto phys_table_ids = get_physical_table_inputs(&ra);
-  executor_->setCatalog(cat_);
   executor_->setDatabaseId(db_id_);
```
Comment (alexbaden) [inline QueryEngine/RelAlgExecutor.cpp:120]: I think there is a better way to do this; consider eliminating state entirely from the Executor and instead pass the "configuration state" for the query down as a single object. Yes, the signatures get bigger for some modules, but setting state like this is error prone (I chased down a few self created bugs during this refactoring).

**Comment by alexbaden** [comment]
I have changed SchemaProvider ID back to 0 (as was intended) and used `getDatabaseId()` instead.

Need to finish sanity_test but ExecuteTest passes, so this is ready for review. If anyone merges, please squash.

**Comment by alexbaden** [comment]
Still working through a few test failures.

**Comment by alexbaden** [comment]
Should pass all tests now.

**Review by ienkovich — APPROVED** [review]
The patch is good! Thanks!

**Pull request merged**